compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp

```diff
@@ -0,0 +1,390 @@
+    batchInitShape[idx.index()] = outputShape[idx.value()];
+  }
+  Value batchEmpty = rewriter.create<tensor::EmptyOp>(loc, batchInitShape,
+                                                      outType.getElementType());
+  Value zero =
+      outType.getElementType().isa<IntegerType>()
+          ? rewriter.create<arith::ConstantOp>(
+                loc, rewriter.getIntegerAttr(outType.getElementType(), 0))
+          : rewriter.create<arith::ConstantOp>(
+                loc, rewriter.getFloatAttr(outType.getElementType(), 0.0));
+  Value batchInit =
+      rewriter.create<linalg::FillOp>(loc, zero, batchEmpty).getResult(0);
+  OpTy batchOp = rewriter.create<OpTy>(
+      loc, TypeRange{batchInit.getType()},
+      ValueRange{extendedLhs.value(), extendedRhs.value()},
+      ValueRange{batchInit});
+  Value newOut = transposeTensor(loc, rewriter, batchOp.getResult(0), outPerm);
+  rewriter.replaceOp(genericOp, newOut);
+  return success();
+}
+
+static LogicalResult
+raiseToBatchVecmat(PatternRewriter &rewriter, linalg::GenericOp genericOp,
+                   linalg::ContractionDimensions contractionDims) {
+  assert(contractionDims.batch.size() == 1 && contractionDims.m.size() == 0 &&
+         contractionDims.n.size() == 1 && contractionDims.k.size() == 1 &&
+         "expected batch vecmat contraction dims");
+  AffineMap vecMap =
+      genericOp.getMatchingIndexingMap(genericOp.getDpsInputOperand(0));
+  AffineMap matMap =
+      genericOp.getMatchingIndexingMap(genericOp.getDpsInputOperand(1));
+  AffineMap outMap =
+      genericOp.getMatchingIndexingMap(genericOp.getDpsInitOperand(0));
+  assert(vecMap.getNumResults() == 2 && matMap.getNumResults() == 3 &&
+         outMap.getNumResults() == 2 && "wrong numResults for indexing maps");
+
+  // Permutation from GenericOp lhs shape to BatchVecmatOp lhs shape
+  SmallVector<int64_t> vecPerm{
+      *(vecMap.getResultPosition(
+          rewriter.getAffineDimExpr(contractionDims.batch[0]))),
+      *(vecMap.getResultPosition(
+          rewriter.getAffineDimExpr(contractionDims.k[0])))};
+  // Permutation from GenericOp rhs shape to BatchVecmatOp rhs shape
+  SmallVector<int64_t> matPerm{
+      *(matMap.getResultPosition(
+          rewriter.getAffineDimExpr(contractionDims.batch[0]))),
+      *(matMap.getResultPosition(
+          rewriter.getAffineDimExpr(contractionDims.k[0]))),
+      *(matMap.getResultPosition(
+          rewriter.getAffineDimExpr(contractionDims.n[0])))};
+  // Permutation from BatchVecmatOp result shape to GenericOp result shape
+  SmallVector<int64_t> outPerm{
+      *(outMap.getResultPosition(
+          rewriter.getAffineDimExpr(contractionDims.batch[0]))),
+      *(outMap.getResultPosition(
+          rewriter.getAffineDimExpr(contractionDims.n[0])))};
+  outPerm = getInverseTransposePerm(outPerm);
+  return liftGenericOp<linalg::BatchVecmatOp>(rewriter, genericOp, vecPerm,
+                                              matPerm, outPerm);
```

Comment:
@bjacob For BatchVecmat, there are no BatchVecmatTransposeA/B ops, but for BatchMatmul would it be better to add a check here for whether the batch dimension is transposed (and then create the different flavors of BatchMatmulTransposeA/B accordingly)? Or is it okay to leave it as a Transpose + BatchMatmul?